CDP4Composition/Reporting/CategoryHierarchy.cs

```diff
@@ -0,0 +1,97 @@
+// <copyright file="CategoryHierarchy.cs" company="RHEA System S.A.">
+//    Copyright (c) 2015-2020 RHEA System S.A.
+//
+//    Author: Sam Gerené, Alex Vorobiev, Alexander van Delft, Cozmin Velciu, Adrian Chivu
+//
+//    This file is part of CDP4-IME Community Edition.
+//    The CDP4-IME Community Edition is the RHEA Concurrent Design Desktop Application and Excel Integration
+//    compliant with ECSS-E-TM-10-25 Annex A and Annex C.
+//
+//    The CDP4-IME Community Edition is free software; you can redistribute it and/or
+//    modify it under the terms of the GNU Affero General Public
+//    License as published by the Free Software Foundation; either
+//    version 3 of the License, or any later version.
+//
+//    The CDP4-IME Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program. If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace CDP4Composition.Reporting
+{
+    using CDP4Common.SiteDirectoryData;
+    using CDP4Common.EngineeringModelData;
+
+    using System.Linq;
+
+    public class CategoryHierarchy
+    {
+        internal class Builder
+        {
+            private readonly Iteration iteration;
+
+            private CategoryHierarchy top, current;
+
+            internal Builder(Iteration iteration, string topLevelCategoryShortName)
+            {
+                this.iteration = iteration;
+                this.AddLevel(topLevelCategoryShortName);
+            }
+
+            private Category GetCategoryByShortName(string shortName)
+            {
+                return this.iteration.Cache
+                    .Select(x => x.Value.Value)
+                    .OfType<Category>()
+                    .Single(x => x.ShortName == shortName);
+            }
+
+            public Builder AddLevel(string categoryShortName)
+            {
+                var category = this.GetCategoryByShortName(categoryShortName);
+
+                var newCategoryHierarchy = new CategoryHierarchy(category, this.current);
+
+                if (this.current == null)
+                {
+                    this.top = this.current = newCategoryHierarchy;
+                }
+                else
+                {
+                    this.current.Child = newCategoryHierarchy;
+                    this.current = this.current.Child;
+                }
+
+                return this;
+            }
+
+            public CategoryHierarchy Build()
+            {
+                return this.top;
+            }
+        }
+
+        #region Hierarchy
```

Comment:
please do not make use of regions, see https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/blob/master/.github/CONTRIBUTING.md#style-guidelines